sql/schemachanger: support complex ALTER COLUMN … SET TYPE in DSC

spilchen opened the pull request:
Previously, complex column type changes that required a backfill reverted to the legacy schema changer. This update now enables support for such changes in the declarative schema changer (DSC).

As with the legacy process, a complex type change involves dropping the column being modified and adding a new one with the updated type. A temporary compute expression is used to map the old column to the new one for backfilling. The process includes performing a double backfill for the primary key (PK): first, both the old and new columns are included in the PK, and then a second backfill is done after the old column is removed. This staged approach is required due to a rule that prevents dropping the column’s compute expression before the dependent column is removed, keeping the expression until after the first backfill.

Since the old and new columns coexist temporarily, the old column is renamed to avoid ambiguity in the PK. A transient name is assigned to the dropped column (e.g., <col>_shadow), which is automatically cleaned up later. This ensures clear differentiation during the transition. A new field was also added to ColumnName to restore the original name if the operation is undone, avoiding the use of a placeholder.

However, there are a few limitations with the current implementation:
- No support for columns that already have a compute expression
- No support for columns with ON UPDATE or DEFAULT expressions
- No support for conversions to enum types
- While column order is preserved in the catalog for SELECT * queries, the order within the column family is not maintained, with the new column being added at the end of the internal ordering.
- Some tests, designed for the legacy schema changer, couldn’t easily be adapted for the DSC.

These limitations are planned to be addressed in future changes.

Epic: CRDB-40232
Closes #127014
Release notes: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133047)
<!-- Reviewable:end -->

rharding6373 reviewed (commented):
Since this largely doesn't affect CDC I'm going to remove myself as a reviewer. Left one nit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133047#-:-O9jjenw7j-nsEk9UIVd:b7zirhe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133047#-O9jjQCI39rvvhtYIdcN:-O9jjQCJEcOJS_ccS713:b-2zny4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e794b2c75bc76150857a0ff23f07b542af7270b1/pkg/ccl/changefeedccl/cdcevent/event_test.go#L601)):*
> ```Go
> 	// Use alter column type to force column reordering.
> 	sqlDB.Exec(t, `SET enable_experimental_alter_column_type_general = true`)
> 	// TODO(spilchen): force the legacy schema changer. When run with the DSC,
> ```

nit: use `TODO(#133040)`


<!-- Sent from Reviewable.io -->

annrpom commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go:
```diff
@@ -228,11 +240,130 @@ func handleGeneralColumnConversion(
+	// Add the new column. It will be identical to the column it is replacing,
+	// except the type will differ, and it will have a transient computed expression.
+	// This expression will reference the original column to facilitate the backfill.
+	spec := addColumnSpec{
```
but the old column is still there until the first backfill what do you think if we say "set target status of old column to drop" and then in the new column text perhaps something like "after this backfill, the old column actually gets dropped"

i do think this is implied with your explanation further down so perhaps this is just a nit

annrpom commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_name.go:
```diff
@@ -26,15 +26,21 @@ func init() {
 						ColumnID: this.ColumnID,
 						Name:     tabledesc.ColumnNamePlaceholder(this.ColumnID),
 					}
+					// If a name was provided for the transition to absent, override the placeholder.
```
for my understanding: is this just setting the new column to the column column's name?

annrpom commented on pkg/sql/schemachanger/scplan/internal/rules/current/dep_alter_column_type.go:
```diff
@@ -52,4 +52,21 @@ func init() {
 			}
 		},
 	)
+
```
ah i guess i had the same question about deprules as you did now :-); do we want to preserve 24.3's rules and then have this be in "current"?

annrpom reviewed (commented):
this is exciting! nice work; i just left some questions for my understanding

also:

> The process includes performing a double backfill for the primary key (PK): first, both the old and new columns are included in the PK

could you elaborate why we need to include these columns in the PK?

let me ask the question about deprules in our -eng channel and then i'll stamp

annrpom reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

also: bazel extended ci failures seem related

spilchen committed:
sql/schemachanger: support complex ALTER COLUMN … SET TYPE in DSC

Previously, complex column type changes that required a backfill reverted to
the legacy schema changer. This update now enables support for such changes in
the declarative schema changer (DSC).

As with the legacy process, a complex type change involves dropping the column
being modified and adding a new one with the updated type. A temporary compute
expression is used to map the old column to the new one for backfilling. The
process includes performing a double backfill for the primary key (PK): first,
both the old and new columns are included in the PK, and then a second backfill
is done after the old column is removed. This staged approach is required due
to a rule that prevents dropping the column’s compute expression before the
dependent column is removed, keeping the expression until after the first
backfill.

Since the old and new columns coexist temporarily, the old column is renamed to
avoid ambiguity in the PK. A transient name is assigned to the dropped column
(e.g., <col>_shadow), which is automatically cleaned up later. This ensures
clear differentiation during the transition. A new field was also added to
ColumnName to restore the original name if the operation is undone, avoiding
the use of a placeholder.

However, there are a few limitations with the current implementation:
- No support for columns that already have a compute expression
- No support for columns with ON UPDATE or DEFAULT expressions
- No support for conversions to enum types
- While column order is preserved in the catalog for SELECT * queries, the
  order within the column family is not maintained, with the new column being
  added at the end of the internal ordering.

Some tests, designed for the legacy schema changer, couldn’t easily be adapted for the DSC.

These limitations are planned to be addressed in future changes.

Epic: CRDB-40232
Closes #127014
Release notes: none

rafiss reviewed (approved):
lgtm! nice work

> Reviewed 43 of 48 files at r1, 1 of 2 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133047#-:-OAJzdAD2oOS64v3rlPt:b-m0nb8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
